----
# **ARTIFICIAL NEURAL NETWORKS**
-----

### OBJECTIVE :

#####  Developing a classification model using Artificial Neural Networks (ANNs) to classify data points from the "Alphabets_data.csv" dataset into predefined categories of alphabets. This exercise aims to deepen your understanding of ANNs and the significant role hyperparameter tuning plays in enhancing model performance


### TASKS :

##### DATA PREPROCESSING :

In [6]:
import warnings
warnings.filterwarnings('ignore')

In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [9]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder ,MinMaxScaler

In [10]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [11]:
from tensorflow.keras.optimizers import Adam, RMSprop

In [12]:
from sklearn.metrics import classification_report

In [13]:
df=pd.read_csv('Alphabets_data.csv')

In [14]:
df

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  \
0          T     2     8      3       5      1     8    13      0      6   
1          I     5    12      3       7      2    10     5      5      4   
2          D     4    11      6       8      6    10     6      2      6   
3          N     7    11      6       6      3     5     9      4      6   
4          G     2     1      3       1      1     8     6      6      6   
...      ...   ...   ...    ...     ...    ...   ...   ...    ...    ...   
19995      D     2     2      3       3      2     7     7      7      6   
19996      C     7    10      8       8      4     4     8      6      9   
19997      T     6     9      6       7      5     6    11      3      7   
19998      S     2     3      4       2      1     8     7      2      6   
19999      A     4     9      6       6      2     9     5      3      1   

       xybar  x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0          6      10       8      0       8      0       8  
1         13       3       9      2       8      4      10  
2         10       3       7      3       7      3       9  
3          4       4      10      6      10      2       8  
4          6       5       9      1       7      5      10  
...      ...     ...     ...    ...     ...    ...     ...  
19995      6       6       4      2       8      3       7  
19996     12       9      13      2       9      3       7  
19997     11       9       5      2      12      2       4  
19998     10       6       8      1       9      5       8  
19999      8       1       8      2       7      2       8  

[20000 rows x 17 columns]

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


In [16]:
df.isnull().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

In [17]:
df.describe()

xbox          ybox         width       height         onpix  \
count  20000.000000  20000.000000  20000.000000  20000.00000  20000.000000   
mean       4.023550      7.035500      5.121850      5.37245      3.505850   
std        1.913212      3.304555      2.014573      2.26139      2.190458   
min        0.000000      0.000000      0.000000      0.00000      0.000000   
25%        3.000000      5.000000      4.000000      4.00000      2.000000   
50%        4.000000      7.000000      5.000000      6.00000      3.000000   
75%        5.000000      9.000000      6.000000      7.00000      5.000000   
max       15.000000     15.000000     15.000000     15.00000     15.000000   

               xbar          ybar         x2bar         y2bar         xybar  \
count  20000.000000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.897600      7.500450      4.628600      5.178650      8.282050   
std        2.026035      2.325354      2.699968      2.380823      2.488475   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.000000      6.000000      3.000000      4.000000      7.000000   
50%        7.000000      7.000000      4.000000      5.000000      8.000000   
75%        8.000000      9.000000      6.000000      7.000000     10.000000   
max       15.000000     15.000000     15.000000     15.000000     15.000000   

            x2ybar        xy2bar         xedge        xedgey         yedge  \
count  20000.00000  20000.000000  20000.000000  20000.000000  20000.000000   
mean       6.45400      7.929000      3.046100      8.338850      3.691750   
std        2.63107      2.080619      2.332541      1.546722      2.567073   
min        0.00000      0.000000      0.000000      0.000000      0.000000   
25%        5.00000      7.000000      1.000000      8.000000      2.000000   
50%        6.00000      8.000000      3.000000      8.000000      3.000000   
75%        8.00000      9.000000      4.000000      9.000000      5.000000   
max       15.00000     15.000000     15.000000     15.000000     15.000000   

            yedgex  
count  20000.00000  
mean       7.80120  
std        1.61747  
min        0.00000  
25%        7.00000  
50%        8.00000  
75%        9.00000  
max       15.00000

#### **MODEL IMPLEMENTATION :**

In [19]:
scaler = StandardScaler()

numeric_features = df.select_dtypes(include=['float64', 'int64']).columns
df[numeric_features] = scaler.fit_transform(df[numeric_features])

print("\nData after Normalization:")
print(df.head())


Data after Normalization:
  letter      xbox      ybox     width    height     onpix      xbar  \
0      T -1.057698  0.291877 -1.053277 -0.164704 -1.144013  0.544130   
1      I  0.510385  1.502358 -1.053277  0.719730 -0.687476  1.531305   
2      D -0.012309  1.199738  0.435910  1.161947  1.138672  1.531305   
3      N  1.555774  1.199738  0.435910  0.277513 -0.230939 -0.936631   
4      G -1.057698 -1.826464 -1.053277 -1.933571 -1.144013  0.544130   

       ybar     x2bar     y2bar     xybar    x2ybar    xy2bar     xedge  \
0  2.365097 -1.714360  0.344994 -0.917071  1.347774  0.034125 -1.305948   
1 -1.075326  0.137561 -0.495072  1.895968 -1.312807  0.514764 -0.448492   
2 -0.645273 -0.973591  0.344994  0.690380 -1.312807 -0.446513 -0.019764   
3  0.644886 -0.232823  0.344994 -1.720796 -0.932724  0.995402  1.266419   
4 -0.645273  0.507945  0.344994 -0.917071 -0.552641  0.514764 -0.877220   

     xedgey     yedge    yedgex  
0 -0.219082 -1.438153  0.122911  
1 -0.219082  0.120081

In [20]:
# One-hot encode the target variable
y = pd.get_dummies(df['letter'])  # Converts the target to one-hot encoding

# Define features and split the dataset
X = df.drop('letter', axis=1)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Construct a basic ANN model with an output layer matching the number of classes
model = Sequential([
    Dense(64, input_shape=(X_train.shape[1],), activation='relu'),  # Input layer and first hidden layer
    Dense(32, activation='relu'),  # Additional hidden layer
    Dense(y_train.shape[1], activation='softmax')  # Output layer with one unit per class
])

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train the model
history = model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

Epoch 1/10
400/400 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.2966 - loss: 2.5721 - val_accuracy: 0.6637 - val_loss: 1.1988
Epoch 2/10
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.6979 - loss: 1.0571 - val_accuracy: 0.7506 - val_loss: 0.8955
Epoch 3/10
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7635 - loss: 0.8188 - val_accuracy: 0.7872 - val_loss: 0.7640
Epoch 4/10
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7983 - loss: 0.7049 - val_accuracy: 0.8109 - val_loss: 0.6618
Epoch 5/10
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8190 - loss: 0.6083 - val_accuracy: 0.8250 - val_loss: 0.5950
Epoch 6/10
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8447 - loss: 0.5212 - val_accuracy: 0.8447 - val_loss: 0.5374
Epoch 7/10
400/400 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.8604 - loss: 0.4797 - val_accuracy: 0.8600 - val_loss: 0.4872
Epoch 8/10
400/400 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8731 - loss: 0.4380 - val_accuracy: 0.

In [21]:
# Evaluate model
test_loss, test_accuracy = model.evaluate(X_test, y_test)
print(f"Test Accuracy: {test_accuracy * 100:.2f}%")

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8882 - loss: 0.3818   
Test Accuracy: 88.40%


In [22]:
# Function to create and compile a model
def create_model(optimizer, activation, neurons, input_shape, num_classes):
    model = Sequential([
        Dense(neurons, input_shape=(input_shape,), activation=activation),
        Dense(neurons // 2, activation=activation),
        Dense(num_classes, activation='softmax')
    ])
    model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
    return model

#### **HYPERPARAMETER TUNING :**

In [23]:
# Define hyperparameters grid
optimizers = ['adam', 'rmsprop']
activations = ['relu', 'tanh']
neurons_list = [64, 128]
batch_sizes = [32, 64]
epochs_list = [10, 20]

In [24]:
# Track the best configuration
best_accuracy = 0
best_params = {}

# Loop through each combination of hyperparameters
for optimizer in optimizers:
    for activation in activations:
        for neurons in neurons_list:
            for batch_size in batch_sizes:
                for epochs in epochs_list:
                    # Create and train the model
                    model = create_model(optimizer=optimizer, activation=activation,
                                         neurons=neurons, input_shape=X_train.shape[1],
                                         num_classes=y_train.shape[1])
                    
                    model.fit(X_train, y_train, epochs=epochs, batch_size=batch_size, verbose=0)
                    
                    # Evaluate the model
                    y_pred = np.argmax(model.predict(X_test), axis=1)
                    y_test_labels = np.argmax(y_test.values, axis=1)
                    accuracy = accuracy_score(y_test_labels, y_pred)
                    
                    # Update best configuration if current is better
                    if accuracy > best_accuracy:
                        best_accuracy = accuracy
                        best_params = {
                            'optimizer': optimizer,
                            'activation': activation,
                            'neurons': neurons,
                            'batch_size': batch_size,
                            'epochs': epochs
                        }

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 720us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 720us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 743us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 876us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 745us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 717us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 734us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 712us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 896us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 763us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 704us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 808us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 733us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 999us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 770us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 917us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 720us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 725us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 735us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 741us/step
125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 999us/step 
125/125 ━━━━━━

#### **EVALUATION**

In [25]:
# Print the best hyperparameters and accuracy
print(f"Best Accuracy: {best_accuracy * 100:.2f}%")
print("Best Hyperparameters:", best_params)

Best Accuracy: 94.92%
Best Hyperparameters: {'optimizer': 'adam', 'activation': 'tanh', 'neurons': 128, 'batch_size': 32, 'epochs': 20}


In [26]:
# Make predictions on the test set
y_pred = model.predict(X_test)
y_pred_classes = y_pred.argmax(axis=1)
y_true = y_test.values.argmax(axis=1)

# Display classification report
print("Classification Report:")
print(classification_report(y_true, y_pred_classes))


125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 813us/step
Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.96      0.96       149
           1       0.93      0.90      0.91       153
           2       0.95      0.89      0.92       137
           3       0.90      0.91      0.91       156
           4       0.92      0.94      0.93       141
           5       0.94      0.94      0.94       140
           6       0.96      0.93      0.95       160
           7       0.90      0.85      0.87       144
           8       0.98      0.91      0.94       146
           9       0.94      0.94      0.94       149
          10       0.84      0.92      0.88       130
          11       0.97      0.95      0.96       155
          12       0.98      0.98      0.98       168
          13       0.93      0.94      0.94       151
          14       0.93      0.94      0.93       145
          15       0.98      0.93      0.95       173
          16   

#### **EVALUATION CRITERIA**

1. **Overall Accuracy**: The model achieved an accuracy of **94%**, indicating strong performance.

2. **Precision**: Most classes have high precision (≥ 0.90), meaning few false positives.

3. **Recall**: High recall (≥ 0.90 for many classes) suggests the model effectively identifies true positives.

4. **F1-Score**: Balanced F1-scores (around 0.90 for many classes) indicate a good trade-off between precision and recall.

5. **Support**: The number of instances varies per class, impacting precision and recall calculations.

6. **Classes Needing Improvement**: 
   - Class `1`: Precision of **0.90** and recall of **0.89**.
   - Class `17`: Precision of **0.85** and recall of **0.93**.

7. **Macro and Weighted Averages**: Both averages are **0.94**, showing consistent performance across classes, accounting for support.

8. **Next Steps**: Consider analyzing the confusion matrix for insights on specific misclassifications.